examples/vinput.c

```diff
@@ -0,0 +1,399 @@
+                              const char *buf, size_t len)
+{
+    int err;
+    unsigned long id;
+    struct vinput *vinput;
+
+    err = kstrtol(buf, 10, &id);
+    if (err) {
+        err = -EINVAL;
+        goto failed;
+    }
+
+    vinput = vinput_get_vdevice_by_id(id);
+    if (IS_ERR(vinput)) {
+        pr_err("vinput: No such vinput device %ld\n", id);
+        err = PTR_ERR(vinput);
+        goto failed;
+    }
+
+    vinput_unregister_vdevice(vinput);
+    device_unregister(&vinput->dev);
+
+    return len;
+failed:
+    return err;
+}
+/* This macro generates class_attr_unexport structure and unexport_store() */
+static CLASS_ATTR_WO(unexport);
+
+static struct attribute *vinput_class_attrs[] = {
+    &class_attr_export.attr,
+    &class_attr_unexport.attr,
+    NULL,
+};
+
+/* This macro generates vinput_class_groups structure */
+ATTRIBUTE_GROUPS(vinput_class);
+
+static struct class vinput_class = {
+    .name = "vinput",
+    .owner = THIS_MODULE,
+    .class_groups = vinput_class_groups,
+};
+
+int vinput_register(struct vinput_device *dev)
+{
+    spin_lock(&vinput_lock);
+    list_add(&dev->list, &vinput_devices);
+    spin_unlock(&vinput_lock);
+
+    pr_info("vinput: registered new virtual input device '%s'\n", dev->name);
+
+    return 0;
+}
+EXPORT_SYMBOL(vinput_register);
+
+void vinput_unregister(struct vinput_device *dev)
+{
+    struct list_head *curr, *next;
+
+    /* Remove from the list first */
+    spin_lock(&vinput_lock);
+    list_del(&dev->list);
+    spin_unlock(&vinput_lock);
+
+    /* unregister all devices of this type */
+    list_for_each_safe (curr, next, &vinput_vdevices) {
+        struct vinput *vinput = list_entry(curr, struct vinput, list);
+        if (vinput && vinput->type == dev) {
+            vinput_unregister_vdevice(vinput);
+            device_unregister(&vinput->dev);
+        }
+    }
+
+    pr_info("vinput: unregistered virtual input device '%s'\n", dev->name);
+}
+EXPORT_SYMBOL(vinput_unregister);
+
+static int __init vinput_init(void)
+{
+    int err = 0;
+
+    pr_info("vinput: Loading virtual input driver\n");
+
+    vinput_dev = register_chrdev(0, DRIVER_NAME, &vinput_fops);
+    if (vinput_dev < 0) {
+        pr_err("vinput: Unable to allocate char dev region\n");
+        goto failed_alloc;
+    }
+
+    spin_lock_init(&vinput_lock);
+
+    err = class_register(&vinput_class);
+    if (err < 0) {
+        pr_err("vinput: Unable to register vinput class\n");
+        goto failed_class;
+    }
+
+    return 0;
+failed_class:
+    class_unregister(&vinput_class);
+failed_alloc:
+    return err;
+}
+
+static void __exit vinput_end(void)
+{
+    pr_info("vinput: Unloading virtual input driver\n");
+
+    unregister_chrdev(vinput_dev, DRIVER_NAME);
+    class_unregister(&vinput_class);
+}
+
+module_init(vinput_init);
+module_exit(vinput_end);
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Tristan Lelong <tristan.lelong@blunderer.org>");
```

Comment:
Remove the name from the source code, and then you should mention Tristan Lelong in the section "Acknowledgements."